### ISSUE TITLE
Blank node skolemization

### ISSUE THREAD

**Issue by mielvds** [LinkedDataFragments/Server.Java#6]
HDT files with blank nodes generate invalid turtle because of blank nodes. Can we create a skolemized Model somehow?

**Comment by RubenVerborgh** [comment]
On the one hand, this seems to be a problem with the HDT library not correctly converting blank nodes to the corresponding Jena representation. On the other hand, the TPF spec clearly says that [components must not be blank nodes](http://www.hydra-cg.com/spec/latest/triple-pattern-fragments/#definition), so we should indeed skolemize them in any case, like the JavaScript implementation does.

**Comment by mielvds** [comment]
[This](http://grepcode.com/file/repo1.maven.org/maven2/fr.inria.eventcloud/eventcloud-api/1.2.1/fr/inria/eventcloud/api/Skolemizator.java) might help, although reprocessing all nodes doesn't seem very performant.

**Comment by RubenVerborgh** [comment]
We should be able to do the same as in the [JavaScript code](https://github.com/LinkedDataFragments/Server.js/blob/v1.1.6/lib/datasources/Datasource.js#L79-L84) by just changing [this function](https://github.com/LinkedDataFragments/Server.Java/blob/master/src/org/linkeddatafragments/datasource/HdtDataSource.java#L107-L113) (probably on the base class level even).

**Comment by mielvds** [comment]
Right, that would be fairly easy, but it would be datasource specific though... Another option would be to create a decorator for https://jena.apache.org/documentation/javadoc/arq/org/apache/jena/riot/WriterGraphRIOT.html

**Comment by RubenVerborgh** [comment]
Well, in the JavaScript version, it's implemented on the base class, so not source-specific. I still think this is possible. The additional complexity here is that `dictionary.getNode` seems to have a bug so that it does not return blank nodes but IRIs that start with `_:`. The generic solution would be to fix that, and then everything works with a generic base method (or RIOT decorator). But then we have some performance loss, because it would first (incorrectly) convert to IRI, then to blank, then to IRI again. So it might be best to have a one-off solution here for performance.

**Comment by mielvds** [comment]
So we'll have to improve the java HDT code no matter what

**Comment by larsgsvensson** [comment]
Hi all, I just stumbled over this and it still seems to be an issue. I'm working on some other things in Server.java (including support for quad formats) so if you can give me any hints on how to fix this issue, I can give it a try.
Thanks,
Lars

**Comment by larsgsvensson** [comment]
As per https://github.com/comunica/comunica/issues/375 the spec now says that data triples MUST NOT contain blank nodes

**Comment by larsgsvensson** [comment]
Is there such a thing as a conformance test suite for TPF servers?

**Comment by RubenVerborgh** [comment]
Not yet unfortunately, but that would indeed be very nice to have.

**Comment by mielvds** [comment]
Sounds like fun. You can assign it to me, I'll try it as friday afternoon thing
